common/crypto_utils/avalon_crypto_utils/crypto_utility.py

```diff
@@ -83,12 +93,17 @@ def encrypt_data(data, encryption_key, iv=None):
           The default is all zeros.iv must be a unique random number for every
           encryption operation.
     """
-    logger.debug("encrypted_session_key: %s", encryption_key)
-    if iv is not None:
-        encrypted_data = crypto.SKENC_EncryptMessage(encryption_key, iv, data)
-    else:
-        encrypted_data = crypto.SKENC_EncryptMessage(encryption_key, data)
-    return encrypted_data
+    cipher = AES.new(bytes(encryption_key), AES.MODE_GCM, bytes(iv))
+    ciphered_data, tag = cipher.encrypt_and_digest(bytes(data))
+    data_list = []
+
+    for i in range(0, len(ciphered_data)):
+        data_list.append(ciphered_data[i])
+
+    for j in range(0, len(tag)):
+        data_list.append(tag[j])
+
+    return tuple(data_list)
```

Comment:
Avalon C++ openSSL encryption function will generate a random IV (12 bytes) if user doesn't provide IV and prepends IV in encrypted data. Similarly the C++ openSSL decryption routine will assume 12 bytes IV is prepended to encrypted data, if user doesn't provide IV for decryption. The python function for encryption and decryption should support similar functionality, if IV is not provided to `encrypt_data()` function.

1) Below is the encrypted message format if IV is not provided by user
 IV = generate_random_bytes(12)
 encrypted_data = IV (12 bytes) + cipher_text + tag (16 bytes)

2) Below is the encrypted message format if IV is provided by user
encrypted_data =  cipher_text + tag (16 bytes)

Reply:
Below is code snippet
```
 def encrypt_data(self,data_str, session_key, iv=None):
        data_bytes = data_str.encode("utf-8")       
        # Generate a random iv
        if iv is None:
            iv = get_random_bytes(12)
            generate_iv = True
        else:
            generate_iv = False

        cipher_aes = AES.new(session_key, AES.MODE_GCM, iv)
        ciphertext, tag = cipher_aes.encrypt_and_digest(data_bytes)

        if generate_iv:
            # if iv passed by user is None, random iv generated
            # above is prepended in encrypted data
            # iv + Cipher + Tag
            result = iv + ciphertext + tag
        else:
            # Cipher + Tag
            result = ciphertext + tag
        return result
```